### ISSUE TITLE
Re-Add Missing Classes for JS-API and migrate to GraalVM-JS

### ISSUE THREAD

**Issue by perkinslr** [RPTools/maptool#2519]
**Is your feature request related to a problem? Please describe.**
Commit 73fa8c89443cc58992d702d0b51f6608eea8ebba removed several unused classes, but caught `net.rptools.maptool.client.script.javascript.api.*` in the crossfire.  This broke the Nashorn-based javascript macro system.  

Additionally, Nashorn itself is going away (removed in OpenJDK JDK 15), so simply re-adding the removed classes would be a temporary solution at best.  There is an alternative Javascript implementation, available from OpenJDK 11 on, called GraalVM-JS, which is generally superior to Nashorn, and can easily be used in Maptool in its place.

**Describe the solution you'd like**
Switch from Nashorn to GraalVM-JS, and re-add the missing interface/wrapper classes, updated for use with the new JS engine.

**Describe alternatives you've considered**
Just re-add the old classes.  They would work without modification, at least until MT switches to JDK 15 (which will likely be soon due to unrelated problems).
Just switch to GraalVM-JS, and remove the built-in reference to `MTScript` which causes `js.eval` to immediately fail.  This would be functional, but essentially leave the javascript only able to do basic math/string manipulation.  Without the API wrapper classes, the map, tokens, drawings, and even client information is inaccessible to JS, and the JS side cannot post messages or roll dice.

**Comment by Phergus** [comment]
Documentation of the `js.eval` function and the JS-API need to be added to the wiki.

**Comment by cwisniew** [comment]
@perkinslr 
This has introduced quite a lit of incompatibilities. Do you have any updates for this? Otherwise I think I am going to revert to my original decision about js.eval() and remove it from 1.9 (to be replaced with something better in a later version).

The move to java 16 has some urgency as it solves some problems people have not even being able to start MapTool so I don't really want to hold up 1.9 because of this.

**Comment by perkinslr** [comment]
There are three related issues here, and a few options for how to best resolve them.  

`clientInfo.faceEdge`, as well as `faceVertex`, `portraitSize`, and `showStatSheet` did not get annotated as `@HostAccess.Export`, so attempts to access them will fail.  This is trivially fixed (and I'll take a look for other un-annotated methods which should be accessible).

The second issue is accessing arrays and maps from javascript.  With `nashorn`, `List` and arrays mapped directly to the javascript `Array`, and `Map` to `object`, with full access to the contents of the `List`, array, or `Map`.  Graalvm-js has this disabled by default.  Instead, read-only access to `List` is possible, by iterating over the list.  I created a wrapper class (`JSList`), which allows similar access to arrays.  From the documentation on `ProxyArray` (the interface `JSList` implements), I believed it would allow read and write access to the array, including via the usual javascript (`myArray[5] = 6`, et cetera) methods.  Since my old code all involved iterating everything, I did not notice this issue in my testing.  There are three options for resolving this:  

1. Slacken the security on arrays, lists and maps *slightly*, allowing full read-write access to them from javascript.  This should be fine, as we only ever *should* be passing **copies** of lists/maps to the javascript side, but if we ever passed a raw list it would be a problem.
1. Improve the wrapper class to allow `[]` operator access to its contents.  This has the advantages of the first option, while not allowing lists to slip through uncoppied.  Any which are not explicitly copied are read-only.  Unfortunately, it's also the most complicated solution to implement.
1. Update the documentation to reflect the read-only, iterate-only status of lists passed from java -> javascript.  If you decide the increased security from never allowing write-access to unproxied arrays is necessary, **and** it is impossible to get the `[]` working properly on proxied arrays, this would be the remaining "fix".  It could easily be accompanied by a utility function on the javascript side which turns java array/list/maps into their javascript counterparts by iterating the Java type and building the javascript type.

The last issue is the type coersion going from javascript to java.  In the nashorn version, arrays and objects returned from javascript got silently converted to strings by json encoding them.  Given that maptool's macro language's internal representation of json objects is as a string, this worked fine.  The challenge is this conversion was done on the Java side previously.  Graalvm-js *does* automatically convert the javascript objects to strings for us, but the strings are *not* valid json.  It *works* if you call `JSON.stringify` when returning complex types from javascript, but it would be better if this were done automatically (but without calling `stringify` on numeric types`).  I'll see if there is an easy way to do it from the Java side, otherwise it can be done in the anonymous context.

**Comment by perkinslr** [comment]
> Documentation of the `js.eval` function and the JS-API need to be added to the wiki.

If I can get an account on the wiki (I assume you mean wiki.rptools.info), I'll be happy to create or update the documentation as needed.  I'm poking at it now.

**Comment by perkinslr** [comment]
Performance-wise, it all should be roughly the same.  The java code is pretty much branchless, and thin, so I expect the JIT will turn it into the same machine code in either case.  (In the JIT-warmup phase, there is one extra tiny indirection in option 2 relative to option 1, but even that is insignificant).  On the subject of json.get, with an index, it is certainly possible to add that API here, which would probably be superior to option 3.

Any objects in a list can only be modified if they carry the `@HostAccess.Export` annotation (and then only through the methods or attributes which are so annotated).  Other Java types are opaque from Javascript, the only thing you can do with them (beyond storing or passing them to another function) is obtain their string representation.  Currently, the only types with the `@HostAccess.Export` annotation are type-specific wrappers (e.g. `JSAPIToken`, which wraps `Token`).  The issue with someone potentially getting access to a raw Java List is they could potentially break it, by duplicating an entry or similar (`mylist.push(mylist[0])` from the javascript side).

**Comment by Azhrei** [comment]
> On the subject of json.get, with an index, it is certainly possible to add that API here, which would probably be superior to option 3.

Oh, I wasn't proposing that API — I hate it and would rather use the brackets (option 2)!  I was just musing that since so many people _have_ used it in their MTscript, they likely wouldn't be thrown off if that same API was used in the initial implementation (with a potential implementation that included the brackets down the road).

**Comment by cwisniew** [comment]
@perkinslr Please disregard @Azhrei comments about json.get(), it's clear he has neither read the comment outlining the problem or understood the issue or he would have known discussing json.get() option was just going to lead you astray.

Since js.eval() will be replaced with something else at some point, I am happier to say it is no longer available than have large breaking changes that people keep filing bug reports against. Minor little things we can live with not large ones though.

Both objects and lists really need to work how they did before access wise.

**Comment by wolph42** [comment]
@cwisniew 
Olla senor, regardless of the fact whether you're right or wrong, I do think that it would be good that when you voice your view on the matter you do that in a less scary way. 'Scary' as in, with that tone you scare off people here. Granted, Ahz is an ancient rock (like you...and me) but he's not the only one active here. I know you like to be short and to the point, but on a public website like this it really gives off a wrong vibe which I know is not your intention. 
Perhaps you can consider revising your wording in the former comment and perhaps elaborate in a bit more respectful way your view on the matter.

Cheers mate!

**Comment by perkinslr** [comment]
So I finally had some time to dig into this.  The list access is trivially solved.  I'll give them a once-over to make sure no uncopied lists get passed through and call that part good enough.

The return type coersion has proven to be a bit tricky.  If I do no further type coersion, `[r: js.eval("return 5")+6]` works as expected, javascript hands the number `5` back.  If I change the closure to `return JSON.stringify((...)())`, lists and objects correctly get encoded, but string literals come out escaped, and numbers come out as strings.  Likewise if I blindly encode the object on the Java side.  The previous implementation used `MacroJavaScriptBridge.JavaScriptToMTScriptType` to do the encoding, which should be fairly easy.  The problem is the Graalvm ScriptEngine compatibility/legacy implementation casts the return from `engine.eval` from the `Value` type, which is public, to the underlying storage type, which is *not* public.  It is easy enough (but still stupidly fiddly) to convert *back* to the `Value` type, but converting to the `Value` type when the return was a string or number again leads to double encoding.  

Anyway, I got it working, but it's messy, and likely slow.  At the least, I can clean it up, but I think it's time to ditch the ScriptEngine implementation.  As a side-effect of poking around the ScriptEngine implementation, I think I have a handle on how modern graalvm works.  For something as simple as what maptool does, it should be roughly the same number of lines of code for the engine and context, and about half the complexity on serializing return values.  I'll give that a spin tomorrow and see about making a PR.

**Comment by Phergus** [comment]
Thanks for digging into this!

**Comment by perkinslr** [comment]
Alright, one more update before the PR...  Dropping ScriptEngine was also quite easy, and makes getting the polymorphic `Value` type trivial.  I've managed to cut the code-duplication for serializing the data type down considerably, but am still not happy with it.  The fundamental problem is the way a direct return is handled is different from the way a returned object in a container is handled, which means two copies of the "what type of thing is this" code, one for top level returns, and one for container returns.

Just to make it clear what I mean, it's the difference between
```
[r: js.eval("return MapTool")]
vs
[r: js.eval("return [MapTool]")]
```
or even
```
[r: js.eval("return 5")]
vs
[r: js.eval("return [5,6,7]")]
``` 

I still have a copy of 1.4.7 running around, so I fired it up long enough to see how different return values were previously serialized.  From the results, and the code, it's pretty clear there wasn't any intentionality behind the previous behavior, but I've written a description of it below, and something of a retro-spec to make the new implementation match, complete with a note of differences.  If there are any changes in how values are serialized, now would be the time to make them, so this is a bit of a request-for-comment.

## Old Behavior
This section includes the input, sample output, and comentary on how the output is generated.

1. Raw numbers
```[r: js.eval("return 5")+6] --> 11```
The `5` is returned from `js.eval` as a `BigDecimal`, which then adds as a number to `6`.
2. Raw strings
```[r: js.eval("return 'a' ")] --> a```
Another trivial case, but note `a` does *not* have quotes around it.
3. Opaque types
```[r: js.eval("return MapTool")] --> net.rptools.maptool.client.script.javascript.api.JSAPIMapTool@35c8bb22```
This is the first case that starts to get interesting.  Returning an instance of something gives the simple string representation of it.  In the case of a `Token` (really a `JSAPIToken` wrapper), it looks like `Token(id=000000003FE0D1721F00000000000000)`
4. Lists created in Java
```[r: js.eval("return MapTool.tokens.getMapTokens()")] -> [Token(id=000000003FE0D1721F00000000000000)]```
5. Items pulled out of lists
```[r: js.eval("return MapTool.tokens.getMapTokens()[0]")] -> Token(id=000000003FE0D1721F00000000000000)```
Pretty simple so far, here's the first brain twister
6. Lists created in javascript
```[r: js.eval("return [MapTool.tokens.getMapTokens()[0]]")] -> InvocationTargetException```
`getMapTokens` returns a `List<Object>`, which converts to a string by converting each of its contents to a string.  If, however, we put the same `Token` into a javascript object, or javascript array, then it tries to JSON serialize the object, which *usually* fails.
7. Opaque types included in lists
```[r: js.eval("return [MapTool]")] -> [{"chat":{},"clientInfo":{},"tokens":{"mapTokens":[]}}]``` if the map is empty
```[r: js.eval("return [MapTool]")] -> InvocationTargetException``` if the map contains a token
The interesting thing to note is that the top level name `MapTool` *is* json serializable, but it tries to serialize its tokens, which are not json serializable.
8. Primitive types in an array 
```[r: js.eval("return [4,5,6]")] -> [4,5,6]``` Note the lack of spaces
```[r: js.eval("return MTScript.getMTScriptCallingArgs()",4,5,6)] -> [4, 5, 6]```
```[r: js.eval("return args",4,5,6)] -> [4, 5, 6]``` the return of `MTScript.getMTScriptCallingArgs()` is bound to the local `args` in the `js.eval` macro
```[r: js.eval("return ['a','b','c']")] -> ["a","b","c"]``` Note the lack of spaces, and the `"`
```[r: js.eval("return args",'a','b','c')] -> [a, b, c]```
9. Primitive types in an `object`

**Comment by perkinslr** [comment]
Alright, one more update before the PR...  Dropping ScriptEngine was also quite easy, and makes getting the polymorphic `Value` type trivial.  I've managed to cut the code-duplication for serializing the data type down considerably, but am still not happy with it.  The fundamental problem is the way a direct return is handled is different from the way a returned object in a container is handled, which means two copies of the "what type of thing is this" code, one for top level returns, and one for container returns.

Just to make it clear what I mean, it's the difference between
```
[r: js.eval("return MapTool")]
vs
[r: js.eval("return [MapTool]")]
```
or even
```
[r: js.eval("return 5")]
vs
[r: js.eval("return [5,6,7]")]
``` 

I still have a copy of 1.4.7 running around, so I fired it up long enough to see how different return values were previously serialized.  From the results, and the code, it's pretty clear there wasn't any intentionality behind the previous behavior, but I've written a description of it below, and something of a retro-spec to make the new implementation match, complete with a note of differences.  If there are any changes in how values are serialized, now would be the time to make them, so this is a bit of a request-for-comment.

## Old Behavior
This section includes the input, sample output, and comentary on how the output is generated.

1. Raw numbers
```[r: js.eval("return 5")+6] --> 11```
The `5` is returned from `js.eval` as a `BigDecimal`, which then adds as a number to `6`.
2. Raw strings
```[r: js.eval("return 'a' ")] --> a```
Another trivial case, but note `a` does *not* have quotes around it.
3. Opaque types
```[r: js.eval("return MapTool")] --> net.rptools.maptool.client.script.javascript.api.JSAPIMapTool@35c8bb22```
This is the first case that starts to get interesting.  Returning an instance of something gives the simple string representation of it.  In the case of a `Token` (really a `JSAPIToken` wrapper), it looks like `Token(id=000000003FE0D1721F00000000000000)`
4. Lists created in Java
```[r: js.eval("return MapTool.tokens.getMapTokens()")] -> [Token(id=000000003FE0D1721F00000000000000)]```
5. Items pulled out of lists
```[r: js.eval("return MapTool.tokens.getMapTokens()[0]")] -> Token(id=000000003FE0D1721F00000000000000)```
Pretty simple so far, here's the first brain twister
6. Lists created in javascript
```[r: js.eval("return [MapTool.tokens.getMapTokens()[0]]")] -> InvocationTargetException```
`getMapTokens` returns a `List<Object>`, which converts to a string by converting each of its contents to a string.  If, however, we put the same `Token` into a javascript object, or javascript array, then it tries to JSON serialize the object, which *usually* fails.
7. Opaque types included in lists
```[r: js.eval("return [MapTool]")] -> [{"chat":{},"clientInfo":{},"tokens":{"mapTokens":[]}}]``` if the map is empty
Note that the keys coorespond to the public members of the class.
```[r: js.eval("return [MapTool]")] -> InvocationTargetException``` if the map contains a token
The interesting thing to note is that the top level name `MapTool` *is* json serializable, but it tries to serialize its tokens, which are not json serializable.
8. Primitive types in an array 
```[r: js.eval("return [4,5,6]")] -> [4,5,6]``` Note the lack of spaces
```[r: js.eval("return MTScript.getMTScriptCallingArgs()",4,5,6)] -> [4, 5, 6]```
```[r: js.eval("return args",4,5,6)] -> [4, 5, 6]``` the return of `MTScript.getMTScriptCallingArgs()` is bound to the local `args` in the `js.eval` macro
```[r: js.eval("return ['a','b','c']")] -> ["a","b","c"]``` Note the lack of spaces, and the `"`
```[r: js.eval("return args",'a','b','c')] -> [a, b, c]```
9. Primitive types in an `object`
```[r: js.eval("return MapTool.clientInfo.libraryTokens()")] -> {Lib:abc=unknown}```
```[r: js.eval("var x = {'Lib:abc': MapTool.clientInfo.libraryTokens()['Lib:abc']}; return x")] -> {"Lib:abc":"unknown"}```
Once again, javascript created objects get an extra level of escaping relative to HashMaps passed in from Java and back out.
10. json interaction on the Java side
```[r: json.get(js.eval('return {"a":5}'), 'a')] -> 5```


## New behavior
These are the considerations in writing the new serializer, in order of importance.
1. Don't break existing non-trivial scripts. 
2. Parsing the return value with the `json.get` method needs to be possible where it intuitively makes sense.
3. Array-like objects should serialize the same regardless of origin, likewise for HashMap/object-like objects
4. Opaque types should remain opaque

To be clear, by non-trivial I mean I don't care if a script which only works on an empty map breaks.  If someone is doing something dumb like converting the `args` array to a string and then splitting that on `, `, I care, but I don't think that is worth breaking number 3 over.  As a corollary to number 1, scripts which previously resulted in errors can have their behavior changed. 
 Numbers 2 and 4 are related, and where I could most benefit from input.  It's worth noting that the `json.get` macro is fairly forgiving, so `[r: json.get("{a:5}", "a")]`,  which *should* choke (as bare words are not valid in json), doesn't.  Likewise, you can use numeric keys.  While the key can be bare, the value cannot be, so `[r: json.get("{a:a}", "a")]` errors.  

Anyway, the first decision is whether to make array-likes serialize following the trampolined-java object's style, `[a, b, c]` or `[1, 2, 3]`, or following the user-created array style `["a","b","c"]`.  I am discounting the otherwise-better `["a", "b", "c"]`, as it is quite similar in how to parse the result to the second option, while potentially breaking existing scripts, and anyone who wants the properly formatted output can use `return JSON.stringify(...` to get it.  That said, I'm inclined to use option 2, as any user-created return arrays will be in that format.  Given that the `ArrayList`s the user can get can also be gotten from macro calls, I'm going to posit that there are no non-trivial scripts which rely on their weird serialization.  They *do* get returned on occasion during script writing, but I very much doubt anyone wrote any parsers for their return.

Likewise for objects/HasMaps.  

Currently, `Token`s can be returned via an `ArrayList`, but not in a user-defined list.  They should be returnable via a user-defined list, but how to do that is not clear.  The option of serializing them out, like `return [MapTool]` currently does is pretty worthless.  We could return them as bare strings (`[Token(id=...), Token(id=...)]`), or we could return *just* their string IDs, as strings, (`["0000000B...", "0000000B"]`), or IDs as bare strings, or...?  
A case could be made to use the first option, as that is what currently happens when bouncing an `ArrayList` from java, but actually parsing that result is essentially useless.  Again, someone who wants the IDs could create the list of IDs in javascript before returning it.

Anyway, here's the spec as it stands.
1. javascript `null` and `undefined` map to the empty string.
2. Top-level primitive types are converted to their MTScript equivalent type.  Numbers to BigDecimal, Strings to Strings.  Neither are escaped
3. Top-level opaque types are converted to their string representation, not escaped, as though the return statement were `return ""+<ret>;`
4. `Array`-likes and `Object`-likes are serialized to a json-like string.  It need not be perfect, but *must* be parsable by the `json.get` macro.  
4a. Items within an `Array`-like or `Object`-like are comma, but not space separated
5. `Array`-like and `Object`-likes contained within the same are recursively processed.
5a. Cycles in the graph must be detected and either raise an error or otherwise not cause a stack overflow.
6. Primitives and opaque types within an `Array`-like or `Object`-like should be suitably escaped, no bare-string values, numbers and booleans as their json representation
7. `null` and `undefined` within an `Array`-like or `Object`-like should map to the json representation of `null`.

## Summary of changes from previous implmentation
1. Some opaque types which previously splatted their public members into an `Array` or `Object` now remain opaque
2. Some opaque types which errored when added to an `Array` or `Object` now remain opaque
3. Java `ArrayList` and `HashMap`s passed into javascript and back now serialize following the same rules as `Array` or `Object`

Number 3 is the one which *might* break some things, but the previous poor behavior is quite unlikely to be used in any real scripts.  Also, given that the previous JS implementation broke nearly a year ago, and went without comment, this is likely the best chance we're going to get to fix these serialization issues.  Yes, `js.eval` will eventually get replaced, but *whatever* replaces it will need to serialize javascript types, so doing that in a sane way *now* will save headache later.

**Comment by Phergus** [comment]
Thanks for the deep dive.  Will look at it more tomorrow when my brain is switched on.

**Comment by Phergus** [comment]
@perkinslr  Status?

**Comment by cwisniew** [comment]
PR #2717 submitted to temporarily disable js.evel() in 1.9 build so that we can get build out the door (there are some who have trouble staring MapTool at the moment that need java 16 bug fixes to be able to do so). This can easily be removed when addressed in later version

**Comment by perkinslr** [comment]
Alright, I'm writing this here so if I get interrupted again it should be faster to pick this up again.  Also, if someone can sanity check me, I wouldn't object.

The hard part is writing the serializer to go from mixed "stuff" to neatly formatted strings.  The heavy lifting can be done via a json library (gson is already used by maptool, and "just works").  But we need to handle cycles, and do some pre-processing before we can ask gson to stringify anything.

When we get an object back from the javascript context, it is wrapped up in a [Value](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/Value.html).  `Value` is a generic wrapper class, which has introspection functions to ask if the contained value is some type or has some property.  It also has conversion functions to return the contained value as a known type.  This lets us handle strings, numbers, and `null` easily.  MapTool types are also easily handled, as the wrappers now all implement an interface which includes a serialization function.  Array-likes (`List`, javascript or Java `Array`, and the like) are also easily identified.  Handling them is simple enough, just convert them to a `List` if needed, sanity check them, and invoke gson.  That just leaves maps (javascript `Object`s, Java `HashMap` and the like).  

### Problem

This is where the fun starts...  The obvious choice is to call `value.hasHashEntries`.  Despite the name, this returns false for both instances of `HashMap` and `PolyglotMap`.  Near as I can tell, it only returns true for custom classes implementing the `ProxyHashMap` interface.  I guess because `PolyglotMap` might use something other than a HashMap internally for some polyglot supported languages.  The second option is to call `value.hasMembers`, but this returns true for objects with attributes annotated by `HostAccess.Export`, so will generate false positives.  The third option is to fetch the underlying object, casting it to `Object`, and then using `instanceof`.  The problem is the underlying storage class `PolyglotMap` is not a public class, so cannot be used via `instanceof`.  This leaves us calling `obj.getClass().getName()`, which is far from ideal.  

### Solution

Here is the solution I think will work.  Primitive types and arrays are already filtered out.  As are MapTool's internal types.  This just leaves `HashMap`s, javascript `Object`s, and any unfiltered opaque Java object the user managed to get.  `Value` has a method, `isHostObject`, which returns true if the object did not originate in the guest language.  If this is false, and `hasMembers` is true, we are dealing with a `PolyglotMap`.  Otherwise we can cast the object to `Object` and use `instanceof` to check if it implements the `Map` interface.

This gives an overall conversion procedure which is reasonably simple.

1. Process primitive types.
2. Process `MapToolJSAPIInterface` implementers (MapTool's internal types)
3. Process arrays, lists, et cetera
4. Check if the object is *not* a `HostObject`.
5. Check if it has members (is this needed if it is not a HostObject?  I don't think so...).
6. If 4 and 5, process `PolyglotMap`s
7. Process HashMaps
8. Process generic opaque types

**Comment by perkinslr** [comment]
Well that was easier to write than I anticipated.  All the examples on the original forum post work, including the ones in the third post here which previously failed.  All the examples from my tests with MT 1.4 produce compatible output (minor differences with spaces or quotes, but json conformant now).  

Cycles in the returned types show up as the string `"[[...]]"`.  Inspired by Python's handling of the same.  It would be easy to throw an exception instead, but making the exception include which object is part of a cycle is not easy.

I will see about opening a pull request in the morning.

**Comment by cwisniew** [comment]
Great work!

 I think it's  fine for the cycles to appear as "[[...]]", It's something that should never occur moving from MTScript to JS (and if it does it means a MT bug needs to be fixed rather than user error to stop it). And from the other way around MT script has no concept of references or pointers so it couldn't handle it anyway. So as long as we document it then its all good.

**Comment by perkinslr** [comment]
I think that is ready for review.  

>  It's something that should never occur moving from MTScript to JS 
Correct.  I cannot even think of a reason we would pass an Array or Map which allows un-specialized Arrays or Maps in to javascript.  Without doing that, a cycle going from MT->javascript is impossible.  

Going the other way, there is likely no "production" use case for it.  But while a user is working on their data structure, they may want to dump it to the chat box for easy viewing.  Not having to dump it in sections or otherwise avoid cycles makes that early development work somewhat easier.  We should make sure to document what `"[[...]]"` means.  I'm also open to using some other sigil string to denote a cycle (by their nature, denoting *which* objects form the cycle is not easy though).

**Comment by Phergus** [comment]
@perkinslr I don't see an account for you on the wiki so we need to get that done.  You can DM me or @melek on Discord with an email address to be used and we'll add an account for you.  Thanks!

**Comment by perkinslr** [comment]
Done.  Currently, there doesn't appear to be a wiki page for the main-thread javascript (no hit on `js.eval`, `javascript` only comes up w.r.t. overlays/frames or on a page that is due for removal).

I am happy to create a new page, but am uncertain what to title it, and where all it should get links pointing to it.

**Comment by Phergus** [comment]
As this is about the macro function `js.eval` we can start with a macro function page by giving it the macro function category it will automatically show up on the macro functions page.  We'll treat this like the various json functions and use the function name as the page name `json.eval`.

**Comment by Phergus** [comment]
Testing.

All of these throw a `org.graalvm.polyglot.PolyglotException`
```
[r: js.eval("MTScript.raiseError('What an Error!?!?')")]<br>
[r: js.eval("MTScript.abort()")]<br>
[r: js.eval("MTScript.mtsAssert(true, 'nothing to see here')")]<br>
[r: js.eval("MTScript.mtsAssert(false, 'something to see here')")]
```

Something that will be useful to point out for users is that `evalMacro()` in this usage will perform the assignment to the MTScript variable `b`  but `execMacro()` does not.
```
[h: a = 5][h: c = 0]
[r: js.eval("MTScript.evalMacro('{b = a + 2d6}')")]
[r: js.eval("MTScript.execMacro('{a = 5d6}')")]
[r: js.eval("MTScript.execMacro('{c = 8d4}')")]
a:[r: a] b:[r: b] c:[r: c]
```
`a` always resolves to 5 and `c` always resolves to 0 but `b` gets the correct values between 7 and 17.
if `c` isn't defined up front as above then MT prompts for a value for `c` when attempting to output it.

**Comment by perkinslr** [comment]
`[r: js.eval("MTScript.mtsAssert(true, 'nothing to see here')")]<br>` does not throw an error for me.  
The rest do, which is what I would expect (though the internal error type is lost).  I assume you bring this up because macros throwing `AssertFunction.AssertFunctionException` and similar have some special handling elsewhere which `PolyglotException` is munging?

`execMacro` provides a new, nested, scope.  So the third line *does* assign to `a`, but unless the value gets used before the `execMacro` call finishes, it gets discarded without effect.  

```
[h: a = 5][h: c = 0]
[h: js.eval("MTScript.evalMacro('{b = a + 2d6}')")]
[h: js.eval("MTScript.execMacro('{a = 5d6}')")]
[h: js.eval("MTScript.execMacro(`
{c = 100}
[h: broadcast('c is '+c)]
`)")]
a:[r: a] b:[r: b] c:[r: c]
```

**Comment by Phergus** [comment]
You are correct.  `[r: js.eval("MTScript.mtsAssert(true, 'nothing to see here')")]` - outputs nothing and does not throw an exception

Under 1.4.1.8, no exceptions:
`[r: js.eval("MTScript.raiseError('What an Error!?!?')")]` - outputs the `'What an Error!?!?` sting
`[r: js.eval("MTScript.abort()")]` - outputs nothing
`[r: js.eval("MTScript.mtsAssert(false, 'something to see here')")]` - outputs nothing

Under current dev code this is the result of the other three which doesn't seem particularly useful.
```
org.graalvm.polyglot.PolyglotException: What an Error!?!? error executing expression js.eval("MTScript.raiseError('What an Error!?!?')").
Error trace : chat
org.graalvm.polyglot.PolyglotException: macro.function.abortFunction.message error executing expression js.eval("MTScript.abort()").
Error trace : chat
org.graalvm.polyglot.PolyglotException: something to see here error executing expression js.eval("MTScript.mtsAssert(false, 'something to see here')").
Error trace : chat
```

**Comment by perkinslr** [comment]
The question is what the purpose of any of them really is...  I assume the `MTScript.abort()` is akin to the `abort()` macro, which should silently halt execution if the predicate is false?  `MTScript.mtsAssert()` should behave like the `assert()` macro, printing a message and halting if the predicate is false?  

What purpose could `MTScript.raiseError()` serve that wouldn't better be served by just running `throw SomeException` from the javascript code?

It should be pretty simple to filter the exception from the `js.eval()` macro, and re-raise it as the underlying `AssertFunction.AssertFunctionException` or `AbortFunction.AbortFunctionException` as appropriate.  I'd argue the `raiseError` should just get removed.  If anything is *actually* using it, they'll get a different error in its place, but not having it will keep anyone from thinking it does something fancier than it does.

**Comment by Phergus** [comment]
> The question is what the purpose of any of them really is...

This is an excellent question but not one I can answer definitively.   My guess is that the MTScript `abort()` and `mtsAssert()` calls should behave like the macro functions as you said.

No idea what the thinking was behind `raiseError()`.  @cwisniew will need to chime in on that.

**Comment by cwisniew** [comment]
`MTScript.abort()` generates an `AbortFunction.AbortFunctionException`
`MTScript.mtsAssert()` generates an `AssertFunction.AssertFunctionException`
Which are errors tested for and raised by the macro writer, at a later stage the ability to capture them was also added  

`MTScript.raiseError()` raises a `ParserException` which can't be trapped by players MTScript Macros.

Since the ability to catch abort and/or assert has been added I can image that someone might be using all three methods.

**Comment by perkinslr** [comment]
And by raising a `ParserException`, the player gets whatever message was passed in to it *without* the full error message (no `org.graalvm.polyglot.PolyglotException: `).  Fair enough.

Solving all three is done in the same place, anyway.  When `catch`ing the exceptions off `evalAnonymous`, we just check if it's a `PolyglotException`, and if it is, if the underlying exception is a `ParserException`.  If it is, re-raise the `ParserException`.

I did notice that the `assert()` macro takes an optional third parameter.  If it is `true`, the error message is wrapped in a localized string.  So I added support for that to `MTScript.mtsAssert()` at the same time.  

There is still one minor issue with `MTScript.raiseError()`.  `ParserException`s bubbling to the top cause the localized string `msg.error.trace` to be printed to the chat log.  This was not the case in 1.4.7.  So the macro 

`[r: js.eval("MTScript.raiseError('a')")]` 

produces the output 

```
a
Error trace : chat
```

Given that `AbortFunctionException` and the like do *not* produce this extra error trace, it should be possible to remove it from `raiseError`, but I don't think it can be done by throwing a raw `ParserException`.

**Comment by Phergus** [comment]
Either way is fine.  Not anticipating having a new release for a bit.

**Comment by perkinslr** [comment]
Well it was easy to find...  Line 462 in `MacroLinkFunction` is where `AbortFunctionException` and `AssertFunctionException` get special handling.  Should be easy to add some sort of javascript exception and treat it the same as the `AssertFunctionException`.  I'm not sure what to call it, or where to put it (in the `MacroJavaScriptBridge`?).  Also, given that it will require a new subclass `ParserException`, do we want to let players trap it somehow?

**Comment by cwisniew** [comment]
I think you can catch and rethrow the correct type in `MacroJavaScriptBridge` it is probably also worth adding the ability to catch an error thrown by `js.eval(...raiseError())`.

**Comment by perkinslr** [comment]
> I think you can catch and rethrow the correct type in MacroJavaScriptBridge

Aye, that part is easy, as the Polyglot exception holds the original exception.  If we want the user to be able to cache exceptions from `raiseError`, we need a new subclass of `ParserException`.  That is also easy, but I don't know if I should make it an inner class in `MacroJavaScriptBridge`, similar to the `AbortFunctionException`, or if it should get stored elsewhere.

**Comment by cwisniew** [comment]
My preference is that its outside of the `MacroJavaScriptBridge` I am not enamoured with the other two being inner classes (oh boy I hope it wasnt me who originally put them there, that would be embarrassing). If you also get feel the need to refactor the other two out from being inner classes I wont be sad about it :)

**Comment by perkinslr** [comment]
According to git blame, it was Azhrei, but since that commit was part of the spotless stuff, I don't think we can blame him.  Anyway, while poking around I found that the `ReturnFunction` also uses an inner class.  With no obvious place to put the new `JavascriptFunctionException`, I created a new package under the `client/functions` package and refactored all the function-specific exceptions to use it.  I hope something like that was what you had in mind.

If you want them someplace else, moving them again should be pretty trivial.

**Comment by Phergus** [comment]
Testing recent fixes under current dev code:
`[r: js.eval("MTScript.raiseError('What an Error!?!?')")]` - outputs the `'What an Error!?!?` sting
`[r: js.eval("MTScript.abort()")]` - outputs nothing
`[r: js.eval("MTScript.mtsAssert(false, 'something to see here')")]` - outputs `something to see here`
No exceptions thrown.  Assuming those are the correct results

These no longer work with current dev code but did under MT 1.5.1:
`[r: js.eval("return MapTool.tokens.getMapTokens()")]`
`[r:js.eval("return MapTool.tokens.getMapTokens()[1].getName()")]` - cuz no tokens are returned

Note that under MT 1.5.1 `getMapTokens()` returned something like this:
`[Token(id=540E696869A14A5A9013093E7171E18C), Token(id=A10F7A578B264C399E0D362724C1C9FE)]`

With current dev code, it returns an empty array [].

`MapTool.getSelectedTokens()` returns an array of the IDs as Strings.
` [ "540E696869A14A5A9013093E7171E18C", "A10F7A578B264C399E0D362724C1C9FE" ]`
Which means doing something like `MapTool.getSelectedTokens()[1].getName()` doesn't work.

This code is 2.5 to 10 times slower on current Dev versus 1.5.1:
```
[h: start = json.get(getInfo("client"), "timeInMs")]
[h: loops = js.eval('var a= 0; for (var i = 0; i < 10000; i++) var a = a +1; return a')]
[h: end = json.get(getInfo("client"), "timeInMs")]
[r: loops]:[r: time = end - start] ms
```

**Comment by Phergus** [comment]
Basic documentation for JavaScript in MapTool in place:
https://wiki.rptools.info/index.php/Category:Javascript_Function

**Comment by perkinslr** [comment]
> [r: js.eval("return MapTool.tokens.getMapTokens()")]

`js.eval` is never trusted (maybe it should be if directly entered by a GM?).  `MapTool.tokens.*` only return tokens owned by the current player if run from an untrusted context.

> `MapTool.getSelectedTokens()` returns an array of the IDs as Strings.

Per 

> This I am ok if the JavaScript opaque token types are serialised as just their id as strings, as this for all intents and purposes is the MTScript version of an opaque token type.

the string representation of the `JSAPIToken` type is its string ID.  So while the array returned from `MapTool.getSelectedTokens()` *looks* like an array of strings (and is converted *to* an array of strings if it is handed back to MTScript, it is actually the token wrappers.  `[r: js.eval("return MapTool.getSelectedTokens()[0].getName()")]` gives the name of the first selected token.

> This code is 2.5 to 10 times slower on current Dev versus 1.5.1:

Per #2975 , `js.eval` now creates a brand new anonymous namespace for each invocation.  This namespace creation has a small overhead.  If you replace `js.eval('` with `js.evalNS('test', '`, the first invocation will pay that overhead (39ms on my machine).  Subsequent calls to the same namespace are much faster (5ms).

This is reflected in the wiki, but a page on migration from pre-1.8 to the new API would be a good idea.

**Comment by Phergus** [comment]
> `js.eval` is never trusted (maybe it should be if directly entered by a GM?). 

Yes, any function directly entered by a GM should behave as if in a trusted context.

> `MapTool.tokens.*` only return tokens owned by the current player if run from an untrusted context.

That doesn't jive with the documentation that says that `MapTool.tokens.getMapTokens()`, "Returns a list of all tokens on all maps."  Nothing is said about 'owned' tokens.  Also, I would expect a function named `getMapTokens` to just return those on the current map or a named map.

**Comment by perkinslr** [comment]
> Yes, any function directly entered by a GM should behave as if in a trusted context.

Noted.

> That doesn't jive with the documentation that says that `MapTool.tokens.getMapTokens()`, "Returns a list of all tokens on all maps." Nothing is said about 'owned' tokens. Also, I would expect a function named `getMapTokens` to just return those on the current map or a named map.

That wiki page was written before untrusted javascript was enabled.  As for the function name, it is odd.  I left it alone to preserve compatibility with pre-1.8 javascript.  Since we've made breaking changes anyway, I think it would make sense to make `MapTool.tokens.getMapTokens()` behave as its name would imply, and add a few other functions to it as well.  And then make the wiki match the actual behavior.

**Comment by cwisniew** [comment]
> > Yes, any function directly entered by a GM should behave as if in a trusted context.
> 
> Noted.
> 
> > That doesn't jive with the documentation that says that `MapTool.tokens.getMapTokens()`, "Returns a list of all tokens on all maps." Nothing is said about 'owned' tokens. Also, I would expect a function named `getMapTokens` to just return those on the current map or a named map.
> 
> Looks like it does only give tokens on the current map.  Not sure why I thought it gave all tokens.
> 
> That wiki page was written before untrusted javascript was enabled.  As for the function name, it is odd.  I left it alone to preserve compatibility with pre-1.8 javascript.  Since we've made breaking changes anyway, I think it would make sense to make `MapTool.tokens.getMapTokens()` behave as its name would imply, and add a few other functions to it as well.  And then make the wiki match the actual behavior, not just for this set, but for all the functions which behave differently in untrusted JS contexts.
> 
> I'm thinking the following functions.
> 
> 1. `getAllTokens` - All tokens accessible to the caller, including unowned tokens which are visible to the caller.

This might be problematic as by necessity all calls on non owned tokens must fail in non trusted connect, so they would end up just being a nuisance that always has to be filtered out.

> 2. `getMapTokens` - All tokens accessible to the caller on either the current map, or the map of a given name.

See above, unless accessible means owned by

> 3. `getOwnedTokens` - All tokens owned by the caller, or owned by the given player name.
> 4. `getTokenByName` - First token with specified name, accessible to the caller.

See above, unless accessible means owned by

> 5. `getLibToken` - First lib token with
 specified name.  Throw an exception if not in trusted namespace.
> 
> Also, the `MapTool.getSelected` and related functions should really be `MapTool.tokens.getSelected`.  

I think the whole API needs some thought put into it. What's currently existing is likely to be deprecated but hang around for comparability reasons

**Comment by cwisniew** [comment]
> As I was implementing a few of these to test things out I got the strong impression that it needs some serious thought. Unfortunately, now (while the nashorn->graalvm migration and the javascript namespaces require some changes anyway) is likely the only chance we'll have for a long time to actually _remove_ or rename anything. I think we should keep the token-finding API tiny until we have a clear picture of what should exist. So the question is what is the minimal set of token-finding functions that are sufficiently complete.

Pretty much all of the existing functions/methods will be deprecated as now we have JavaScript running in two different threads and one of those threads (WebView) doesn't have access to the data, and since unfortunately both threads are GUI threads we can't get around this with synchronized blocks as we will end up with at best times of slow UI response but also likely to cause deadlocks at some stage. This means the majority of the JavaScript API will have to be based on promises. So anything added to the MapTool.* objects for graalvm would only be short term (a couple of releases) before we tell people not to use it

> 
> If I expand the `getMapTokens` to accept a parameter for the map name, that would allow finding any token anywhere, and is reasonably likely to make the cut for the final API. It's also about the most minimal change.

See above :) It will all be deprecated :(

> 
> As a separate issue, do we move the `MapTool.get*` functions? Do we redirect them to `MapTool.tokens.get*`, knowing that we'll never actually remove the old functions? Or do we leave them in the wrong place as an anachronism of the earlier implementation?
> 
> My preference would be to move them now, while the 2 groups actually using `js.*` are migrating scripts anyway, and before new frameworks come out using them, but I'm not really picky.

If you want to break it now is the time to do it... But keep in mind what I said about it all going to be deprecated anyway

**Comment by Phergus** [comment]
The issues I reported above have been resolved with the recent PR.  I did notice that there is a *lot* of log output.  It seems that every `js.eval()` command run produces this:

```
[To redirect Truffle log output to a file use one of the following options:
* '--log.file=<path>' if the option is passed using a guest language launcher.
* '-Dpolyglot.log.file=<path>' if the option is passed using the host Java launcher.
* Configure logging using the polyglot embedding API.]
[engine] WARNING: The polyglot context is using an implementation that does not support runtime compilation.
The guest application code will therefore be executed in interpreted mode only.
Execution only in interpreted mode will strongly impact the guest application performance.
For more information on using GraalVM see https://www.graalvm.org/java/quickstart/.
To disable this warning the '--engine.WarnInterpreterOnly=false' option or use the '-Dpolyglot.engine.WarnInterpreterOnly=false' system property.
[AWT-EventQueue-0] INFO org.reflections.Reflections - Reflections took 9 ms to scan 1 urls, producing 5 keys and 12 values 
```
Potential for huge logs I imagine.

Not sure if it was always producing the warning about running interpreted instead of using runtime compilation.
